neo/SmartContract/InteropService.cs

```diff
@@ -64,6 +67,52 @@ public static partial class InteropService
         public static readonly uint System_Storage_Delete = Register("System.Storage.Delete", Storage_Delete, 0_01000000, TriggerType.Application);
         public static readonly uint System_StorageContext_AsReadOnly = Register("System.StorageContext.AsReadOnly", StorageContext_AsReadOnly, 0_00000400, TriggerType.Application);
 
+        private static bool CheckItemForNotification(StackItem state)
+        {
+            int size = 0;
+            List<StackItem> items_checked = new List<StackItem>();
+            Queue<StackItem> items_unchecked = new Queue<StackItem>();
+            while (true)
+            {
+                switch (state)
+                {
+                    case Struct array:
+                        foreach (StackItem item in array)
+                            items_unchecked.Enqueue(item);
+                        break;
+                    case Array array:
+                        if (items_checked.All(p => !ReferenceEquals(p, array)))
+                        {
+                            items_checked.Add(array);
+                            foreach (StackItem item in array)
+                                items_unchecked.Enqueue(item);
+                        }
+                        break;
+                    case Boolean _:
+                    case ByteArray _:
+                    case Integer _:
+                        size += state.GetByteLength();
+                        break;
+                    case InteropInterface _:
+                        return false;
+                    case Map map:
+                        if (items_checked.All(p => !ReferenceEquals(p, map)))
+                        {
+                            items_checked.Add(map);
+                            foreach (var pair in map)
+                            {
+                                size += pair.Key.GetByteLength();
+                                items_unchecked.Enqueue(pair.Value);
+                            }
+                        }
+                        break;
+                }
+                if (size > MaxNotificationSize) return false;
+                if (items_unchecked.Count == 0) return true;
+                state = items_unchecked.Dequeue();
+            }
+        }
+
```

Comment:
Shouldn't we add this on NeoVM as a general CheckItem method, passing maxsize as parameter?

Reply:
I thought the same, but is not used there...

Reply:
So what does this test test exactly?

Reply:
> Shouldn't we add this on NeoVM as a general CheckItem method, passing maxsize as parameter?

Totally agree!

Potential risks may exist because complex data structures can be build. NeoVM do need such checking in to avoid runtime issue.

Reply:
Checking what? Notification is not in NeoVM layer.